### PR TITLE
Improve asynchronous FL with full FedBuff in the Collab CIFAR-10 example

### PR DESCRIPTION
**Pull request by ZiyueXu77**
## What this changes

- implements online FedBuff with client jobs that remain active across global model versions
- exposes K (active jobs), B (update-buffer size), and O (open-slot dispatch threshold), defaulting to K=8, B=4, O=1
- uses the same 8-client CIFAR-10 setup as the existing FedAvg example: the shared preparation/splitting utilities, ModerateCNN, site shards and transforms, batch size 64, four local epochs, SGD with momentum, cosine LR scheduling, and step-weighted deltas
- defers evaluation until training completes, then evaluates every saved global version on the same server-side CIFAR-10 test set so evaluation cannot perturb asynchronous completion order
- documents the synchronous O=B=K configuration and includes the matched 400-update comparison figure

## Validation

- ran the existing FedAvg baseline for 50 rounds (8 updates/round)
- ran the Collab workflow synchronously for 50 rounds with K=B=O=8; its curve aligns with the existing FedAvg baseline
- ran Collab FedBuff for 100 global versions with K=8, B=4, O=1 (400 accepted updates)
- verified the copied `prepare_data.sh` resolves and runs the shared CIFAR-10 utility directly
- verified the eight shared alpha=0.5/seed-0 shards and exact equality of all 18 ModerateCNN initialization tensors
- verified recipe finalization, Python compilation, Black formatting, shell syntax, and `git diff --check`

### CODE REVIEW

**Comment by ZiyueXu77** [comment]
@nvkevlu, thanks for the question! I think they make perfect sense, and thus extended this example to fully cover the FedBuff algo with configurations that can also include sync scenarios - align with our edge exampe. Please review.

File: examples/advanced/collab/pt_async_cifar10/trainer.py [hunk 1]
```diff
@@ -49,121 +62,190 @@ def __init__(
...
-
-            inputs = inputs.to(device)
-            labels = labels.to(device)
-            optimizer.zero_grad()
-            loss = criterion(model(inputs), labels)
-            loss.backward()
-            optimizer.step()
-            total_loss += loss.item()
+        if self.profile == "fedavg":
+            num_steps = self.aggregation_epochs * len(dataloader)
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/collab/pt_async_cifar10/trainer.py]: Thanks. The latest revision removed arbitrary supplied shards: job.py owns split creation through the shared CIFAR-10 splitter, and the documented eight-client alpha=0.5, seed-0 setup produces non-empty shards. We are keeping additional malformed-data handling out of this limited baseline example.

File: examples/advanced/collab/pt_async_cifar10/trainer.py [hunk 2]
```diff
@@ -49,121 +62,190 @@ def __init__(
...
+        self._scheduler = None
+        self._dataloaders = {}
 
     @collab.init
     def initialize(self):
         torch.set_num_threads(self.num_threads)
         self.data_root = collab.get_app_prop("data_root", self.data_root)
-        self._manifest = load_manifest(self.data_root)
+        if self.profile == "native":
+            self._manifest = load_manifest(self.data_root)
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/collab/pt_async_cifar10/trainer.py]: Resolved structurally in the latest revision: the separate profile and user-supplied train_idx_root path were removed. job.py now always creates the shared CIFAR-10 split path before constructing Cifar10Trainer, so Path(None) is no longer reachable through this example.

File: examples/advanced/collab/pt_async_cifar10/job.py [hunk 3]
```diff
@@ -84,39 +95,46 @@ def make_recipe(args, manifest: dict) -> CollabRecipe:
...
     client = Cifar10Trainer(
         data_root=args.data_root,
         local_batch_size=args.local_batch_size,
         local_iters=args.local_iters,
         local_lr=args.local_lr,
         device=device,
+        profile=args.profile,
+        train_idx_root=args.train_idx_root,
+        aggregation_epochs=args.aggregation_epochs,
+        total_client_rounds=args.total_client_rounds,
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/collab/pt_async_cifar10/job.py:141]: Fixed in the latest revision: make_recipe now rejects aggregation_epochs < 1 and total_client_rounds < 1 before constructing the client and scheduler.

File: examples/advanced/collab/pt_async_cifar10/async_aggregator.py [hunk 4]
```diff
@@ -145,149 +209,276 @@ def _evaluate_model(self, global_model: dict[str, torch.Tensor], round_index: in
...
-        desired_clients = (
-            len(physical_names) if self.clients_per_round is None else min(self.clients_per_round, len(physical_names))
+    def _watch_call_failures(self, results) -> None:
+        # Iteration waits for every outcome from this nonblocking group call.
+        # Successful values were already handled by the response callback.
+        for _ in results:
+            pass
+        for physical_name, error in results.failures.items():
+            self._outcomes.put(_ClientOutcome(physical_name=physical_name, error=error))
+
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/collab/pt_async_cifar10/async_aggregator.py:334]: Thanks. This example intentionally assumes healthy simulator clients and keeps failure recovery out of scope so the FedBuff scheduling path stays focused. I have left the watcher unchanged; production failure-recovery behavior belongs in a broader runtime or example change.

File: examples/advanced/collab/pt_async_cifar10/model.py [hunk 5]
```diff
@@ -21,6 +21,61 @@
 import torch.nn.functional as F
 
 
+# ModerateCNN is the model used by
+# examples/advanced/cifar10/pt/cifar10-sim/cifar10_fedavg. Its architecture is
+# derived from IBM FedMA (MIT license; see that example's src/model.py).
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/collab/pt_async_cifar10/model.py:47]: Fixed in 93945ecf by adding the complete IBM FedMA MIT notice and source attribution to model.py.

File: examples/advanced/collab/pt_async_cifar10/trainer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/collab/pt_async_cifar10/trainer.py]: Fixed in dec05370. Although the shared splitter normally guarantees at least 10 samples per client, the trainer now explicitly rejects a shard with zero training batches before calculating the average loss.

File: examples/advanced/collab/pt_async_cifar10/trainer.py [hunk 6]
```diff
@@ -16,154 +16,181 @@
...
         )
-        self.logger.info(
-            f"[{collab.site_name}] loaded prepared CIFAR-10 metadata for "
-            f"{self._manifest['num_clients']} logical clients"
+        self._local_model = ModerateCNN()
+        self._optimizer = torch.optim.SGD(self._local_model.parameters(), lr=self.local_lr, momentum=0.9)
+        self._scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
+            self._optimizer,
+            T_max=self.total_client_rounds * self.aggregation_epochs,
+            eta_min=self.local_lr * 0.01,
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/advanced/collab/pt_async_cifar10/trainer.py:91]: Fixed in a70cbd2a. The persistent scheduler now stops advancing at T_max, so clients that receive more than --total-client-rounds assignments remain at eta_min instead of entering the rising half of another cosine cycle. The CLI help and README now document this behavior.

**Comment by chesterxgchen** [comment]
@ZiyueXu77 which version this update is targeted to ?

**Comment by ZiyueXu77** [comment]
> @ZiyueXu77 which version this update is targeted to ?

Preferalbly 2.9 to company CollabAPI if time permits. The current example showcased async but not truly fedbuff, so I made this update. But it is just an example so can also go to main/2.10 later.

**Review by nvkevlu — APPROVED** [review]

**Comment by pcnudde** [comment]
Let's put this in 2.10. We can merge it in main after we branch.
